File: js/Letterrooster.js
// -------- LIJST VAN STRINGS --------
class RijstVanStrings {
    roosterVanString(str, rijlen) {
        let ret = [];
        // console.assert(isInt(str.length / rijlen), "Ongeldige string");
        for (let i = 0; i < str.length; i += rijlen) {
            // Stopt als het einde van de string gevonden is met een halve rij
            ret.push(str.slice(i, Math.min(i+rijlen, str.length)));
        }
        return ret;
    }

    vierkantRoosterVanString(str) {
        const rijlen = Math.sqrt(str.length);
        console.assert(isInt(rijlen), "ongeldige string");
        let ret = [];
        for (let i = 0; i < str.length; i += rijlen) {
            ret.push(str.slice(i, i+rijlen));
        }
        return ret;
    }

    // Afgeraden, hier is het beter om een rooster van strings te gebruiken
    setValue(rooster, x, y, waarde) {
        console.assert(x < rooster.length, "Ongeldige X");
        const oudStr = rooster[x];
        console.assert(y < rooster[x].length, "Ongeldige Y");
        const nieuwstr = oudStr.substr(0, y) + waarde + oudStr.substr(y+1);
        rooster[x] = nieuwstr;
    }

    // Afgeraden, hier is het beter om een rooster van strings te gebruiken
    setValueReturn(rooster, x, y, waarde) {
        console.assert(x < rooster.length, "Ongeldige X");
        const oudStr = rooster[x];
        console.assert(y < rooster[x].length, "Ongeldige Y");
        const nieuwstr = oudStr.substr(0, y) + waarde + oudStr.substr(y+1);
        let nieuwRooster = rooster.map(rij => rij.slice());
        nieuwRooster[x] = nieuwstr;
        return nieuwRooster;
    }

    getValue(rooster, x, y) {
        console.assert(x < rooster.length, "Ongeldige X");
        console.assert(y < rooster[x].length, "Ongeldige Y");
        return rooster[x][y];
    }

    toString(rooster) {
        return rooster.join("");
    }
}

// -------- ROOSTER VAN STRINGS --------
class RoosterVanStrings {
    roosterVanString(str, rijlen) {
        let ret = [];
        // console.assert(isInt(str.length / rijlen), "Ongeldige string");
        for (let i = 0; i < str.length; i += rijlen) {
            // Stopt als het einde van de string gevonden is met een halve rij
            ret.push(str.slice(i, Math.min(i+rijlen, str.length)).split(""));
        }
        return ret;
    }

    vierkantRoosterVanString(str) {
        const rijlen = Math.sqrt(str.length);
        console.assert(isInt(rijlen), "ongeldige string");
        let ret = [];
        for (let i = 0; i < str.length; i += rijlen) {
            ret.push(str.slice(i, i+rijlen).split(""));
        }
        return ret;
    }

    // Overschrijft de parameter
    setValue(rooster, x, y, waarde) {
        console.assert(x < rooster.length, "Ongeldige X");
        console.assert(y < rooster[x].length, "Ongeldige Y");
        rooster[x][y] = waarde;
    }

    setValueReturn(rooster, x, y, waarde) {
        console.assert(x < rooster.length, "Ongeldige X");
        console.assert(y < rooster[x].length, "Ongeldige Y");
        let nieuwRooster = rooster.map(rij => rij.slice());
        nieuwRooster[x][y] = waarde;
        return nieuwRooster;
    }

    getValue(rooster, x, y) {
        console.assert(x < rooster.length, "Ongeldige X");
        console.assert(y < rooster[x].length, "Ongeldige Y");
        return rooster[x][y];
    }

    toString(rooster) {
        return rooster.map(rij => rij.join("")).join("");
    }
}

// -------- HULPFUNCTIES --------

function isInt(n) {
    n === parseInt(n, 10);
}

File: Python/Letterrooster.py
#!/usr/bin/env python3 

# -------- LIJST VAN STRINGS --------
class LijstVanStrings(object):
    def roosterVanString(self, tekst, rijlen):
        # assert (len(tekst) / rijlen).is_integer(), "Ongeldige string"
        ret = [tekst[i: min(i+rijlen, len(tekst)) ] for i in range(0, len(tekst), rijlen)]
        return ret
    
    def vierkantRoosterVanString(self, tekst):
        rijlen = len(tekst) ** 0.5
        assert rijlen.is_integer(), "Ongeldige string"
        rijlen = int(rijlen)
        ret = [tekst[i: min(i+rijlen, len(tekst)) ] for i in range(0, len(tekst), rijlen)]
        return ret
    
    # Afgeraden, hier is het beter om een rooster van strings te gebruiken
    def setValue(self, rooster, x, y, waarde):
        assert (x < len(rooster)), "Ongeldige X"
        oudStr = rooster[x]
        assert (y < len(rooster[x])), "Ongeldige Y"
        nieuwStr = oudStr[:y] + waarde + oudStr[y+1:]
        rooster[x] = nieuwStr

    # Afgeraden, hier is het beter om een rooster van strings te gebruiken
    def setValueReturn(self, rooster, x, y, waarde):
        assert (x < len(rooster)), "Ongeldige X"
        oudStr = rooster[x]
        assert (y < len(rooster[x])), "Ongeldige Y"
        nieuwStr = oudStr[:y] + waarde + oudStr[y+1:]
        nieuwRooster = [rij[:] for rij in rooster]
        nieuwRooster[x] = nieuwStr
        return nieuwRooster

    def getValue(self, rooster, x, y):
        assert (x < len(rooster)), "Ongeldige X"
        assert (y < len(rooster[x])), "Ongeldige Y"
        return rooster[x][y]
    
    # Waarschijnlijk te gebruiken als __str__() 
    def toString(self, rooster):
        return "".join(rooster)

# -------- ROOSTER VAN STRINGS --------
class RoosterVanStrings(object):
    def roosterVanString(self, tekst, rijlen):
        # assert (len(tekst) / rijlen).is_integer(), "Ongeldige string"
        ret = [list(tekst[i: min(i+rijlen, len(tekst)) ]) for i in range(0, len(tekst), rijlen)]
        return ret
    
    def vierkantRoosterVanString(self, tekst):
        rijlen = len(tekst) ** 0.5
        assert rijlen.is_integer(), "Ongeldige string"
        rijlen = int(rijlen)
        ret = [list(tekst[i: min(i+rijlen, len(tekst)) ]) for i in range(0, len(tekst), rijlen)]
        return ret
    
    def setValue(self, rooster, x, y, waarde):
        assert (x < len(rooster)), "Ongeldige X"
        assert (y < len(rooster[x])), "Ongeldige Y"
        rooster[x][y] = waarde

    def setValueReturn(self, rooster, x, y, waarde):
        assert (x < len(rooster)), "Ongeldige X"
        assert (y < len(rooster[x])), "Ongeldige Y"
        nieuwRooster = [rij[:] for rij in rooster]
        nieuwRooster[x][y] = waarde
        return nieuwRooster

    def getValue(self, rooster, x, y):
        assert (x < len(rooster)), "Ongeldige X"
        assert (y < len(rooster[x])), "Ongeldige Y"
        return rooster[x][y]
    
    # Waarschijnlijk te gebruiken als __str__() 
    def toString(self, rooster = [""]):
        return "".join(["".join(rij) for rij in rooster])
